fix(conversational_retrieval_chain.ts): keep the chat history context alongside the vector store context

joaopcm opened the pull request:
The `question_generator_template` in `conversational_retrieval_chain.ts` file was updated to provide more clarity and include an optional chat history excerpt. The template now instructs to return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question. The updated template also includes the phrase "Standalone question and optional chat history excerpt:".

<!--
Thank you for contributing to LangChainJS! Your PR will appear in our next release under the title you set above. Please make sure it highlights your valuable contribution.

To help streamline the review process, please make sure you read our contribution guidelines:
https://github.com/hwchase17/langchainjs/blob/main/CONTRIBUTING.md

If you are adding an integration (e.g. a new LLM, vector store, or memory), please also read our additional guidelines for integrations:
https://github.com/hwchase17/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

Replace this block with a description of the change, the issue it fixes (if applicable), and relevant context.

Finally, we'd love to show appreciation for your contribution - if you'd like us to shout you out on Twitter, please also include your handle below!
-->

@jopcmelo

<!-- Remove if not applicable -->

Fixes #1800

joaopcm commented:
CC @nfcampos

jacoblee93 commented:
I like this, just worried about potential regressions for folks using it in production right now. Let me circulate it a bit with the Python crew - maybe adding it under a flag would be a good compromise.

Would also want to add some tests/comparisons/docs if possible but hold off for now.

jacoblee93 commented on langchain/src/chains/chat_vector_db_chain.ts:
```diff
@@ -11,12 +11,12 @@ import { CallbackManagerForChainRun } from "../callbacks/manager.js";
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type LoadValues = Record<string, any>;
 
-const question_generator_template = `Given the following conversation and a follow up question, rephrase the follow up question to be a standalone question.
```
This one is deprecated, no need to modify it

joaopcm replied:
Yes, that makes sense! Let me remove this change from this file

joaopcm replied:
Done ;)

jacoblee93 commented:
Consensus is this is useful but definite worries about regression - let's go with a flag in `fromLLM`. How about `extractConversationContext` in `questionGeneratorChainOptions`?

joaopcm commented:
> Consensus is this is useful but definite worries about regression - let's go with a flag in `fromLLM`. How about `extractConversationContext` in `questionGeneratorChainOptions`?

@jacoblee93, tt really is concerning to add, it might be something close to a breaking change. Let's go with a flag instead! I'll update the PR in a few hours :)

jacoblee93 commented:
> > Consensus is this is useful but definite worries about regression - let's go with a flag in `fromLLM`. How about `extractConversationContext` in `questionGeneratorChainOptions`?
> 
> @jacoblee93, tt really is concerning to add, it might be something close to a breaking change. Let's go with a flag instead! I'll update the PR in a few hours :)

Ok cool! Would be great to have a few integration tests too

joaopcm commented:
@jacoblee93, could you review it again, please?

I just added:
- new tests testing the behavior with and without `questionGeneratorChainOptions.extractConversationContext`
- updated the docs including this new feature

jacoblee93 commented:
Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave an incorrect response:

```
  await chain.call({
    question:
      "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
  });

  const res2 = await chain.call({
    question: "What is the airspeed velocity of an unladen swallow?",
  });
```

GPT-4 would probably help but given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, to start I'd love to use this as an example of how folks can customize the chain in the docs. Does that sound alright to you?

juzarantri commented:
> Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave a confidently incorrect response:
> 
> ```
>   await chain.call({
>     question:
>       "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
>   });
> 
>   const res2 = await chain.call({
>     question: "What is the airspeed velocity of an unladen swallow?",
>   });
>   // It returned "Mitochondria are the powerhouse of the cell"
> ```
> 
> GPT-4 would probably help, but it still breaks the abstraction a bit in that the generated question is also not technically a standalone question anymore.
> 
> Given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, I think there's a more moderate starting point of adding this as an example of how folks can customize the chain in the docs.
> 
> Basically, we'd encourage people to try this and potentially tweak on their own:
> 
> ```
>   const chain = ConversationalRetrievalQAChain.fromLLM(
>     model,
>     vectorStore.asRetriever(),
>     {
>       memory: new BufferMemory({
>         memoryKey: "chat_history",
>         returnMessages: true,
>       }),
>       questionGeneratorChainOptions: {
>         template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
> Chat History:
> {chat_history}
> Follow Up Input: {question}
> 
> Your answer should follow the following format:
> \`\`\`
> Use the following pieces of context to answer the users question.
> If you don't know the answer, just say that you don't know, don't try to make up an answer.
> ----------------
> <Relevant chat history excerpt as context here>
> 
> Standalone question: <Rephrased question here>
> \`\`\`
> 
> Your answer:`
>       },
>     }
>   );
> ```
> 
> Does that sound alright to you?


So how can we maintain the history of each user?
Suppose there are 10 user simultaneously using this and the chain is one so I think all the response would be based on all the chats till now right?

And I guess still this version isn't yet published

its-anas commented:
> Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave a confidently incorrect response:
> 
> 
> 
> ```
> 
>   await chain.call({
> 
>     question:
> 
>       "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
> 
>   });
> 
> 
> 
>   const res2 = await chain.call({
> 
>     question: "What is the airspeed velocity of an unladen swallow?",
> 
>   });
> 
>   // It returned "Mitochondria are the powerhouse of the cell"
> 
> ```
> 
> 
> 
> GPT-4 would probably help, but it still breaks the abstraction a bit in that the generated question is also not technically a standalone question anymore.
> 
> 
> 
> Given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, I think there's a more moderate starting point of adding this as an example of how folks can customize the chain in the docs.
> 
> 
> 
> Basically, we'd encourage people to try this and potentially tweak on their own:
> 
> ```
> 
>   const chain = ConversationalRetrievalQAChain.fromLLM(
> 
>     model,
> 
>     vectorStore.asRetriever(),
> 
>     {
> 
>       memory: new BufferMemory({
> 
>         memoryKey: "chat_history",
> 
>         returnMessages: true,
> 
>       }),
> 
>       questionGeneratorChainOptions: {
> 
>         template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
> 
> 
> Chat History:
> 
> {chat_history}
> 
> Follow Up Input: {question}
> 
> 
> 
> Your answer should follow the following format:
> 
> \`\`\`
> 
> Use the following pieces of context to answer the users question.
> 
> If you don't know the answer, just say that you don't know, don't try to make up an answer.
> 
> ----------------
> 
> <Relevant chat history excerpt as context here>
> 
> 
> 
> Standalone question: <Rephrased question here>
> 
> \`\`\`
> 
> 
> 
> Your answer:`
> 
>       },
> 
>     }
> 
>   );
> 
> ```
> 
> 
> 
> Does that sound alright to you?

Will this really have a working memory? Which args to use with chain.call() then?

joaopcm commented:
@juzarantri 
> So how can we maintain the history of each user?
Suppose there are 10 user simultaneously using this and the chain is one so I think all the response would be based on all the chats till now right?
>
> And I guess still this version isn't yet published

I'm not sure if I understand you. Why would you create only one chat for 10 or more users? In this case, Langchain isn't capable of keeping track of different human messages in its memory since it only stores messages from `human` and `ai`.

If you want to have many users using your chain, you'd have to create different session IDs for each of them so that each of them have their own chat history.

joaopcm commented:
> Have tried it out a bit and I think there's a lot of edge cases to where I'm nervous about adding this new prompt directly into the framework - for example, the model hallucinated a bad standalone question here and gave a confidently incorrect response:
> 
> ```
>   await chain.call({
>     question:
>       "I have a friend called Bob. He's 28 years old. He'd like to know what is the powerhouse of the cell?",
>   });
> 
>   const res2 = await chain.call({
>     question: "What is the airspeed velocity of an unladen swallow?",
>   });
>   // It returned "Mitochondria are the powerhouse of the cell"
> ```
> 
> GPT-4 would probably help, but it still breaks the abstraction a bit in that the generated question is also not technically a standalone question anymore.
> 
> Given we can pass the prompt in as a `template` into the `questionGeneratorChainOptions` directly, I think there's a more moderate starting point of adding this as an example of how folks can customize the chain in the docs.
> 
> Basically, we'd encourage people to try this and potentially tweak on their own:
> 
> ```
>   const chain = ConversationalRetrievalQAChain.fromLLM(
>     model,
>     vectorStore.asRetriever(),
>     {
>       memory: new BufferMemory({
>         memoryKey: "chat_history",
>         returnMessages: true,
>       }),
>       questionGeneratorChainOptions: {
>         template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
> Chat History:
> {chat_history}
> Follow Up Input: {question}
> 
> Your answer should follow the following format:
> \`\`\`
> Use the following pieces of context to answer the users question.
> If you don't know the answer, just say that you don't know, don't try to make up an answer.
> ----------------
> <Relevant chat history excerpt as context here>
> 
> Standalone question: <Rephrased question here>
> \`\`\`
> 
> Your answer:`
>       },
>     }
>   );
> ```
> 
> Does that sound alright to you?

@jacoblee93, that makes sense! This is how I ended up fixing this issue at @getproducthub. Let me update the PR to include only the examples then :)

joaopcm commented:
> Will this really have a working memory? Which args to use with chain.call() then?

@its-anas, it does have a working memory. There's no secret with `chain.call()`. You just have to provide the question to the chain and it will consider relevant character history excerpt alongside the vector store context under the hood.

Let me know if I've answered your question :)

its-anas commented:
> @its-anas, it does have a working memory. There's no secret with `chain.call()`. You just have to provide the question to the chain and it will consider relevant character history excerpt alongside the vector store context under the hood.
> 
> Let me know if I've answered your question :)


@joaopcm based on your explanation and @jacoblee93 code, then the following chain must have the memory working. But when i interact with it, it doesn't remember anything.

Shouldn't i pass chat_history somewhere? When i pass it to chain.call() i get an error. So when i should pass it?

```js
const question = 'Hi, my name is John';

const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

const history = [
  ['Hi, my name is John', 'Hi John! How can I assist you today?'],
];

const chain = ConversationalRetrievalQAChain.fromLLM(
  new ChatOpenAI({
    temperature: 0,
    modelName: 'gpt-3.5-turbo',
  }),
  vectorStore.asRetriever(),
  {
    memory: new BufferMemory({
      memoryKey: 'chat_history',
      returnMessages: true,
    }),
    questionGeneratorChainOptions: {
      template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.

                Chat History:
                {chat_history}
                Follow Up Input: {question}

                Your answer should follow the following format:
                \`\`\`
                Use the following pieces of context to answer the users question.
                If you don't know the answer, just say that you don't know, don't try to make up an answer.
                ----------------
                <Relevant chat history excerpt as context here>

                Standalone question: <Rephrased question here>
                \`\`\`

                Your answer:`,
    },
  },
);

const response = await chain.call({
  question: sanitizedQuestion,
  // chat_history: history, > this result in Error: input values have 2 keys, you must specify an input key or pass only 1 key as input
});

```

joaopcm commented:
> > @its-anas, it does have a working memory. There's no secret with `chain.call()`. You just have to provide the question to the chain and it will consider relevant character history excerpt alongside the vector store context under the hood.
> > Let me know if I've answered your question :)
> 
> @joaopcm based on your explanation and @jacoblee93 code, then the following chain must have the memory working. But when i interact with it, it doesn't remember anything.
> 
> Shouldn't i pass chat_history somewhere? When i pass it to chain.call() i get an error. So when i should pass it?
> 
> ```js
> const question = 'Hi, my name is John';
> 
> const sanitizedQuestion = question.trim().replaceAll('\n', ' ');
> 
> const history = [
>   ['Hi, my name is John', 'Hi John! How can I assist you today?'],
> ];
> 
> const chain = ConversationalRetrievalQAChain.fromLLM(
>   new ChatOpenAI({
>     temperature: 0,
>     modelName: 'gpt-3.5-turbo',
>   }),
>   vectorStore.asRetriever(),
>   {
>     memory: new BufferMemory({
>       memoryKey: 'chat_history',
>       returnMessages: true,
>     }),
>     questionGeneratorChainOptions: {
>       template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.
> 
>                 Chat History:
>                 {chat_history}
>                 Follow Up Input: {question}
> 
>                 Your answer should follow the following format:
>                 \`\`\`
>                 Use the following pieces of context to answer the users question.
>                 If you don't know the answer, just say that you don't know, don't try to make up an answer.
>                 ----------------
>                 <Relevant chat history excerpt as context here>
> 
>                 Standalone question: <Rephrased question here>
>                 \`\`\`
> 
>                 Your answer:`,
>     },
>   },
> );
> 
> const response = await chain.call({
>   question: sanitizedQuestion,
>   // chat_history: history, > this result in Error: input values have 2 keys, you must specify an input key or pass only 1 key as input
> });
> ```

@its-anas, try adding `inputKey` to `BufferMemory`:
```
memory: new BufferMemory({
    memoryKey: 'chat_history',
    inputKey: 'question',
    returnMessages: true,
  }),
```

joaopcm commented:
@its-anas , also, you might want to provide the history messages directly to `BufferMemory` instead of passing it to the `chain.call`.

Something like this:
```ts
const pastMessages = [
  new HumanChatMessage("My name's Jonas"),
  new AIChatMessage("Nice to meet you, Jonas!"),
];

const memory = new BufferMemory({
  chatHistory: new ChatMessageHistory(pastMessages),
});
```

its-anas commented:
@joaopcm Thanks for helping me about this. But  I tried to do it like you said and it's still not remembering the conversation.

Note: when i asked it about what inside the chat history, it answered with what is stored in the vector database instead of what inside chat history.


```js
import { AIChatMessage, HumanChatMessage } from 'langchain/schema';
import { BufferMemory, ChatMessageHistory } from 'langchain/memory';

const question = 'Question should be here';

const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

const history = [
  ['Hi, my name is John', 'Hello John! How can I assist you today?'],
  [
    "I'm from New York, you?",
    'I am an AI assistant and do not have a physical location.',
  ],
  [
    "Ok, but you remember who i am and where i'm from?",
    "I'm sorry, but as an AI assistant, I don't have the ability to remember personal information about individuals.",
  ],
  [
    'But what inside chat_history?',
    "The chat history contains information about the website's pages, such as the contact page.",
  ],
];

const pastMessages = history.reduce(
  (acc, subArray) => [
    ...acc,
    new HumanChatMessage(subArray[0]),
    new AIChatMessage(subArray[1]),
  ],
  [],
);

const model = new ChatOpenAI({
  temperature: 0,
  modelName: 'gpt-3.5-turbo',
});

const chain = ConversationalRetrievalQAChain.fromLLM(
  model,
  vectorstore.asRetriever(),
  {
    memory: new BufferMemory({
      memoryKey: 'chat_history',
      inputKey: 'question',
      chatHistory: new ChatMessageHistory(pastMessages),
      returnMessages: true,
    }),
    questionGeneratorChainOptions: {
      template: `Given the following conversation and a follow up question, return the conversation history excerpt that includes any relevant context to the question if it exists and rephrase the follow up question to be a standalone question.

                Chat History:
                {chat_history}
                Follow Up Input: {question}

                Your answer should follow the following format:
                \`\`\`
                Use the following pieces of context to answer the users question.
                If you don't know the answer, just say that you don't know, don't try to make up an answer.
                ----------------
                <Relevant chat history excerpt as context here>

                Standalone question: <Rephrased question here>
                \`\`\`

                Your answer:`,
    },
  },
);

const response = await chain.call({
  question: sanitizedQuestion,
});

```

jacoblee93 commented:
You'd need to initialize `ChatMessageHistory` like this: 
```
import { BufferMemory, ChatMessageHistory } from "langchain/memory";
import { HumanChatMessage, AIChatMessage } from "langchain/schema";

const pastMessages = [
  new HumanChatMessage("My name's Jonas"),
  new AIChatMessage("Nice to meet you, Jonas!"),
];

const memory = new BufferMemory({
  chatHistory: new ChatMessageHistory(pastMessages),
});
```

Let's move help requests to Discord though.

jacoblee93 commented:
@joaopcm yeah let's just add it to the docs as an example for now - I think this issue might require more structural changes to fix in a robust way.

its-anas commented:
Thanks @joaopcm, i updated Langchain and memory is working. Thanks @jacoblee93 for you answer. Sorry guys for ruining your PR comments.

joaopcm commented:
Sounds good, dude! Thanks!

jacoblee93 commented:
> Sounds good, dude! Thanks!

Thank you for pushing hard on this!

SantoshKumarRavi commented:
yes. solution working fine.
but while streaming chat-history and rephrased question also in the token.
how to stream the output only as token ?

  ```
const model = new ChatOpenAI({
      openAIApiKey: this.config.get('openai.keys'),
      temperature: 0,
      modelName: modelName,
      streaming: true,
      callbacks: [
        {
          handleLLMNewToken(token: string) {
            console.log({ token });
          },
        },
      ],
    });
```

my issue is more or like this 
https://github.com/langchain-ai/langchain/issues/7599
